src/xmipp/libraries/reconstruction/subtract_projection.cpp

```diff
@@ -84,287 +86,283 @@
+	readParticle(row);
+	row.getValueOrDefault(MDL_ANGLE_ROT, part_angles.rot, 0);
+	row.getValueOrDefault(MDL_ANGLE_TILT, part_angles.tilt, 0);
+	row.getValueOrDefault(MDL_ANGLE_PSI, part_angles.psi, 0);
+	roffset.initZeros(2);
+	row.getValueOrDefault(MDL_SHIFT_X, roffset(0), 0);
+	row.getValueOrDefault(MDL_SHIFT_Y, roffset(1), 0);
+	roffset *= -1;
+	projectVolume(*projector, P, sizeImg, sizeImg, part_angles.rot, part_angles.tilt, part_angles.psi, ctfImage);
+	Pctf = applyCTF(row, P);
+	transformerPf.FourierTransform(Pctf(), PFourier, false);
+	transformerIf.FourierTransform(I(), IFourier, false);
+}
 
- MultidimArray<double> computeRadQuotient(MultidimArray<double> &rq, const MultidimArray<double> & rmI,
-		 const MultidimArray<double> &rmP){
-	rq = rmI/rmP;
-	FOR_ALL_ELEMENTS_IN_ARRAY1D(rq) {
-		rq(i) = std::min(rq(i), 1.0);
-	}
-	return rq;
- }
-
- void ProgSubtractProjection::runIteration() {
-	transformer.FourierTransform(P(),PFourier,false);
-	POCSFourierAmplitudeProj(IFourierMag,PFourier, lambda, radQuotient, (int)XSIZE(I()));
-	transformer.inverseFourierTransform();
-	POCSMinMaxProj(P(), Imin, Imax);
-	transformer.FourierTransform();
-	POCSFourierPhaseProj(PFourierPhase,PFourier);
-	transformer.inverseFourierTransform();
- }
+MultidimArray< std::complex<double> > ProgSubtractProjection::computeEstimationImage(const MultidimArray<double> &Img, 
+const MultidimArray<double> &InvM, FourierTransformer &transformerImgiM) {
+	ImgiM().initZeros(Img);
+	FOR_ALL_DIRECT_ELEMENTS_IN_MULTIDIMARRAY(Img)
+		DIRECT_MULTIDIM_ELEM(ImgiM(),n) = DIRECT_MULTIDIM_ELEM(Img,n) * DIRECT_MULTIDIM_ELEM(InvM,n);
+	transformerImgiM.FourierTransform(ImgiM(),ImgiMFourier,false);
+	return ImgiMFourier;
+}
 
- Image<double> ProgSubtractProjection::thresholdMask(Image<double> &m){
- 	double maxMaskVol;
- 	double minMaskVol;
- 	m().computeDoubleMinMax(minMaskVol, maxMaskVol);
-	FOR_ALL_DIRECT_ELEMENTS_IN_MULTIDIMARRAY(m())
-		DIRECT_MULTIDIM_ELEM(m(),n)=(std::abs(DIRECT_MULTIDIM_ELEM(m(),n)>maxMaskVol/20)) ? 1:0;
-	FilterG2.applyMaskSpace(m());
-	FOR_ALL_DIRECT_ELEMENTS_IN_MULTIDIMARRAY(m())
-		DIRECT_MULTIDIM_ELEM(m(),n)=(std::abs(DIRECT_MULTIDIM_ELEM(m(),n)>0.5)) ? 1:0;
-	for (int n=0; n<10; ++n) {
-		closing2D(m(), m(), 4, 0, 16);
+ double ProgSubtractProjection::evaluateFitting(const MultidimArray< std::complex<double> > &y,
+                                                const MultidimArray< std::complex<double> > &yp) const{
+	double sumY = 0;
+	double sumY2 = 0;
+	double sumE2 = 0;
+	FOR_ALL_DIRECT_ELEMENTS_IN_MULTIDIMARRAY(y) {
+		double realyn = real(DIRECT_MULTIDIM_ELEM(y, n)); 
+		double imagyn = imag(DIRECT_MULTIDIM_ELEM(y, n)); 
+		double ereal = realyn - real(DIRECT_MULTIDIM_ELEM(yp, n));
+		double eimag = imagyn - imag(DIRECT_MULTIDIM_ELEM(yp, n));  
+		sumE2 += ereal*ereal + eimag*eimag;
+		sumY += realyn + imagyn;
+		sumY2 += realyn*realyn + imagyn*imagyn;
 	}
- 	return m;
+	auto meanY = sumY / (2.0*(double)MULTIDIM_SIZE(y));
```

Comment:
I don't think so, because `y` is the Fourier transform of the particle and therefore `MULTIDIM_SIZE(y)` is the size of the FT of the particle, so it should never be 0 and if at some (unlikely) case the FT of the particle is 0, the program will fail from the beginning.